### DISCUSSION TITLE
offset vs. scale

### DISCUSSION THREAD

**Discussion by tom-hoffman** [libfive/libfive#534] [General]
I'm a little confused about exactly how offset is being calculated -- and in particular how to describe it to students with a somewhat rudimentary understanding of 3d geometry.  In 2d graphics we've been scaling sprites, so they're used to the idea that a scale factor of two makes a sprite twice as big (doubles both the x and y axis).

offset is clearly meant to do something different... how is it calculated?

**Comment by mkeeter** [comment]
The scale functions are what you're looking for, and should be equivalent to scaling sprites (although they're continuous, rather than discrete, so you won't get the same pixelation that you'd see with sprites).

I'd recommend skipping offset, since there's no good analogy from 2D graphics.

Under the hood, offset is changing the setpoint of the distance field.  For example, a sphere is
```
f(x, y, z) = sqrt(x*x + y*y + z*z) - r
```
where the sphere is solid at points where `f(x, y, z) < 0`.

The offset functions are sliding that setpoint up or down, i.e. doing
```
f'(x, y, z) = (sqrt(x*x + y*y + z*z) - r) + offset
```

The exact behavior depends on the shape of the underlying distance field, so it's not at all equivalent to scaling.

**Reply by tom-hoffman** [reply]
Sorry I'm being such a spaz.  I swear I scanned through the shape reference over and over again and managed to miss scale somehow... I am working on [a reference sheet](https://docs.google.com/document/d/1tUFySvvoVrtRilYvX5myqngzg28NYdc1XdL8wdr5Vaw/edit?usp=sharing) using the contract format we use from Racket, but for that I'm only down to the letter "p" I think!

I'm hoping to use libfive with kids in a month so I'm being less patient than I would be with my regular hobby projects.  Hopefully I'll have some cool projects from the kids to share.

**Reply by lf94** [reply]
(I recently created nodejs bindings, if your students are more familiar with JavaScript. Here's an example: https://github.com/lf94/node-libfive/blob/main/examples/shipping-box.js - but sounds like they're already bitten by the lisp bug ;))

**Comment by omgitsraven** [comment]
The short answer is, the new surface is that many units (inches, whatever) out from where the old surface was.   Like sticking pins into the surface, and then papering over them.  (The slightly more technical answer is that you're offsetting all the values in the distance field by the offset amount, which amounts to basically the same thing as just drawing the boundary at a different value in the field.)
